client/web/src/repo/RepoRevisionSidebarFocusableTree.tsx

```diff
@@ -0,0 +1,24 @@
+import { useEffect, useRef } from 'react'
+
+import { Tree, TreeNode, TreeProps } from '@sourcegraph/wildcard/src'
+
+export interface FocusableTreeProps {
+    focusKey?: string
+}
+
+/**
+ * Renders {@link Tree} and focuses it when {@link focusKey} changes.
+ */
+export function FocusableTree<N extends TreeNode>(props: FocusableTreeProps & TreeProps<N>): JSX.Element {
+    const { focusKey, ...rest } = props
+
+    const ref = useRef<HTMLUListElement>(null)
```

Comment:
```suggestion
    const ref = useRef<TreeRef>(null)
```

How did this not fail? 😱

Reply:
🤷🏻 maybe because `HTMLUListElement ` matches `TreeRef` interface (has `focus` method)?

Reply:
I mean it has `focus` but it also has some properties that only DOM nodes have as well so this should really not work. Maybe the typescript workaround for generic refs doesn't work? :(

Reply:
> Maybe the typescript workaround for generic refs doesn't work?

It seems to work
<img width="477" alt="Screenshot 2023-03-03 at 15 09 44" src="https://user-images.githubusercontent.com/25318659/222728345-106a7561-28ca-43cc-8c34-b0bb0b726f41.png">
<img width="1128" alt="Screenshot 2023-03-03 at 15 09 34" src="https://user-images.githubusercontent.com/25318659/222728353-91ab3a48-b024-4398-8bb5-461decfb7245.png">

But yes, we can be more explicit and define ref type as `TreeRef` 👍🏻

Reply:
https://github.com/sourcegraph/sourcegraph/pull/46829/commits/7851d3d7e82c5d373dad931a9f1da1acad7910d2

Reply:
I’m still confused here. Does this mean `useImperativeHandle` merges the ref with the object you define? e.g. what happens if you do `ref.current.innerHTML = 'foobar'`? This would be valid on a `UList` but not on the `TreeRef` 😅 

Also the error message is confused because we don't want to assign HTMLUListElement ref into TreeRef but instead it should be assigned the other way around haha. Seems like this is a covariant/contravariant issue with your workaround because it looses the nuance that this is one over the other (I don't know exactly which term is the correct one here lol)

https://github.com/microsoft/TypeScript/issues/30748#issuecomment-480197036

Ah the _sound_ type system strikes again 🙃

Reply:
> what happens if you do ref.current.innerHTML = 'foobar'? This would be valid on a UList but not on the TreeRef

You're right! With `TreeRef` as a type we get an expected error
<img width="402" alt="Screenshot 2023-03-03 at 15 31 29" src="https://user-images.githubusercontent.com/25318659/222732988-d66df7ec-6b1a-4b0d-99ed-692b0fe69d8f.png">


> Seems like this is a covariant/contravariant issue with your workaround

Maybe 🤷🏻 
Do you have ideas on how we can define safer types on the `Tree` component with generic type and `forwardRef` combo?

Reply:
> Do you have ideas on how we can define safer types on the Tree component with generic type and forwardRef combo?

no, I did a quick google session but it's not as simple as in Flowtype to define a property as contravariant so I'd say we ignore it for now 😓